File: backend/services/taskService.js
const TaskModel = require('../models/Task');

const getAllTasks = (userId) => {
  return TaskModel.findAllByUser(userId);
};

const getTaskById = (id, userId) => {
  const task = TaskModel.findByIdAndUser(id, userId);
  if (!task) {
    const error = new Error('Task not found');
    error.statusCode = 404;
    throw error;
  }
  return task;
};

const createTask = ({ title, description, status, userId }) => {
  return TaskModel.create({ title, description, status, userId });
};

const updateTask = (id, userId, data) => {
  // Ensure task belongs to user
  const existing = TaskModel.findByIdAndUser(id, userId);
  if (!existing) {
    const error = new Error('Task not found');
    error.statusCode = 404;
    throw error;
  }

  return TaskModel.update(id, userId, data);
};

const deleteTask = (id, userId) => {
  const deleted = TaskModel.delete(id, userId);
  if (!deleted) {
    const error = new Error('Task not found');
    error.statusCode = 404;
    throw error;
  }
};

module.exports = { getAllTasks, getTaskById, createTask, updateTask, deleteTask };

File: taskiflow/lib/features/tasks/services/task_service.dart
// ─────────────────────────────────────────────────────────────────────────────
// task_service.dart
// Service layer untuk semua operasi CRUD task.
// ─────────────────────────────────────────────────────────────────────────────

import '../../../core/constants/api_constants.dart';
import '../../../core/network/api_client.dart';
import '../models/task_model.dart';

class TaskService {
  TaskService._();
  static final TaskService instance = TaskService._();

  final _client = ApiClient.instance;

  // ── Get All Tasks ─────────────────────────────────────────────────────────
  Future<List<TaskModel>> getTasks() async {
    final response = await _client.get(ApiConstants.tasks);
    final data     = response['data'] as Map<String, dynamic>;
    final taskList = data['tasks'] as List<dynamic>;

    return taskList
        .map((json) => TaskModel.fromJson(json as Map<String, dynamic>))
        .toList();
  }

  // ── Get Task By ID ────────────────────────────────────────────────────────
  Future<TaskModel> getTaskById(String id) async {
    final response = await _client.get(ApiConstants.taskById(id));
    return TaskModel.fromJson(
      response['data']['task'] as Map<String, dynamic>,
    );
  }

  // ── Create Task ───────────────────────────────────────────────────────────
  Future<TaskModel> createTask({
    required String title,
    String? description,
    String status = 'pending',
  }) async {
    final response = await _client.post(
      ApiConstants.tasks,
      body: {
        'title'      : title,
        if (description != null && description.isNotEmpty) 'description': description,
        'status'     : status,
      },
    );
    return TaskModel.fromJson(
      response['data']['task'] as Map<String, dynamic>,
    );
  }

  // ── Update Task ───────────────────────────────────────────────────────────
  Future<TaskModel> updateTask(
    String id, {
    String? title,
    String? description,
    String? status,
  }) async {
    final body = <String, dynamic>{};
    if (title != null)       body['title']       = title;
    if (description != null) body['description'] = description;
    if (status != null)      body['status']      = status;

    final response = await _client.put(
      ApiConstants.taskById(id),
      body: body,
    );
    return TaskModel.fromJson(
      response['data']['task'] as Map<String, dynamic>,
    );
  }

  // ── Delete Task ───────────────────────────────────────────────────────────
  Future<void> deleteTask(String id) async {
    await _client.delete(ApiConstants.taskById(id));
  }
}
